text/0000-const-fn-ptr.md

```diff
@@ -0,0 +1,183 @@
+also `const fn`s. Which makes `fn`s and `const fn`s more consistent and
+increases symmetry.
+
+This is a small addition to the language which enhances `const fn`s and allows
+non-`const` `fn`s to restrict what a function taken as a parameter may do.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The term "const function pointer" or "const fn ptr" for short refers to the type
+`const fn(input_types..) -> return_type`. Since it is a type, you may use it
+in a type alias:
+
+```rust
+type MyFn = const fn(u32) -> u32;
+```
+
+This type alias may then be used in a `const fn` to take another `const fn`
+as a parameter:
+
+```rust
+const fn twice(arg: u32, fun: MyFun) -> u32 {
+    fun(fun(arg))
+}
+
+const fn add_one(arg: u32) -> u32 {
+    arg + 1
+}
+
+fn main() {
+    const THREE: u32 = twice(1, add_one);
+    assert_eq!(THREE, 3);
+}
+```
+
+You may also directly write:
+
+```rust
+const fn twice(arg: u32, fun: const fn(u32) -> u32) -> u32 {
+    fun(fun(arg))
+}
+```
+
+Const function pointers are not limited to use in `const fn`s.
+You may also use them for normal `fn`s:
+
+```rust
+fn twice(arg: u32, fun: const fn(u32) -> u32) -> u32 {
+    fun(fun(arg))
+}
+```
+
+Since const function pointers are const, you may also write:
+
+```rust
+fn main() {
+    const PTR: const fn(u32) -> u32 = twice;
+
+    // ..
+}
+```
+
+It is also important to note that the following is legal today:
+
+```rust
+fn main() {
+    let fn_ptr : fn(u32) -> u32 = add_one;
+}
+```
+
+That is: a `const fn` can coerce into a regular `fn` pointer.
+And so with this RFC, the following would be legal:
+
+```rust
+fn main() {
+    let hof_ptr_1 = const fn(u32, const fn(u32) -> u32) -> u32 = twice;
+    let hof_ptr_2 = fn(u32, const fn(u32) -> u32) -> u32 = twice;
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Assuming we currently have (using [LBNF](https://github.com/BNFC/bnfc/blob/master/docs/lbnf.rst) notation):
+
+```
+separator Type ","
+
+FnPtr. Type ::= "fn" "(" [Type] ")" "->" Type ;
+
+-- And other production rules as well...
+```
+
+The following is also allowed (now):
+
+```
+ConstFnPtr. Type ::= "const" "fn" "(" [Type] ")" "->" Type ;
+```
+
+Semantically, only a `const fn fun_name(input_types) -> retr_type {..}` may be
+used where a `const fn(input_types) -> retr_type` type is expected. Which
+type checking (now) ensures. This is done as type checking of normal function
+pointers are done, but also checking that the given `fn` is a `const fn`.
+
+It should be noted that `const fn(.., const fn(x) -> y, ..) -> z` does not
+coerce to `fn(.., fn(x) -> y, ..) -> z` as that would be unsound since
+the outer `fn` may now be given a pointer to a normal `fn` which would execute
+in a `const fn` context thereby subverting the rules of `const fn`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some may argue that we don't need HOFs that are `const fn`.
```

Comment:
Function pointers require some additional checks in miri to ensure noone transmutes function pointers in "bad" ways. It's probably fine to convert a `const fn(&u32)` to a `const fn(Option<&u32>)`, but not the other way around. Additionally someone might convert from `fn()` to `const fn()` and try to call that.

While these things are "forbidden" at runtime (UB), it's not clear how they should be checked at compile-time.

Reply:
Somewhat related: you can use the target's call ABI (or will be able to, soon enough), and if they match up, you can reuse the bits - if register bank/memory indirection doesn't match at all, it's quite UB. But also I mostly care about checking calls, not casts.

Reply:
@oli-obk 

> Additionally someone might convert from `fn()` to `const fn()` and try to call that.

Can't we simply forbid that conversion? Would an alternative be to have UB at compile time (is that possible?) if such a cast is done improperly?

@eddyb Elaborate?

Reply:
> Can't we simply forbid that conversion? Would an alternative be to have UB at compile time (is that possible?) if such a cast is done improperly?

The conversion is not preventable, you can always do `*(&foo as *const Foo as *const Bar)` to hide any conversion. The final call can be checked, but doing that right will require some implementation effort.